N°2293 DBUpdate : save changed fields and corresponding previous values

piRGoif opened the pull request:
In 2.7.0-beta we added DBObject::$m_aChanges
Later on we added the DBObject::ListChangesUpdated as a getter for use in \iApplicationObjectExtension::OnDBUpdate

But that was not enough as we didn't have access to orig values.

We removed this and add in DBObject :
$m_aPreviousValuesForUpdatedAttributes attribute, set at the very beginning of DBUpdate
ListPreviousValuesForUpdatedAttributes() getter

piRGoif committed:
N°2293 DBUpdate : save changed fields and corresponding previous values for callbacks
* update PHPDoc
* remove m_aChanges and ListChangesUpdated() that were introduced in 2.7.0-beta
* add m_aPreviousValuesForUpdatedAttributes and ListPreviousValuesForUpdatedAttributes()

piRGoif committed:
:memo Woops forgot to change one PHPDoc

piRGoif committed:
:memo: Some more PHPDoc O:)

Molkobain commented on core/dbobject.class.php:
```diff
@@ -3305,6 +3308,30 @@ public function DBUpdate()
+			return;
+		}
+
+		$aPreviousValuesForUpdatedAttributes = array_intersect_key($this->m_aOrigValues, $aChanges);
```
Except from the result of that line that needs to be tested with a real use case, sound all good to me! 👌

piRGoif replied:
Christophe did some test, this is validated, I'm closing the conversation :)

bruno-ds commented on core/dbobject.class.php:
```diff
@@ -134,10 +136,11 @@ abstract class DBObject implements iDisplay
 	/**
-	 * @var array attname => currentvalue Persists changes for {@link DBUpdate}
+	 * @var array attname => value : value before the last {@link Set} call. Set at the beginning of {@link DBUpdate}.
+	 * @see ListPreviousValuesForUpdatedAttributes getter for this attribute
```
I think you should type it : 
ListPreviousValuesForUpdatedAttributes::ListPreviousValuesForUpdatedAttributes()

cf : https://github.com/Combodo/iTop/tree/develop/.doc#known-limitations

piRGoif replied:
Wooops thanks, I forgot !
I fixed that for the added @see... and also I changed the @link I did put everywhere to @see, as @link should be used only for URI.
Tell me if those modifications are ok now ?

piRGoif replied:
Saw with Bruno : the rendering is good enough for now. Just a little prb on the way @uses are printed, but we will make a review on the whole doc later on.

bruno-ds commented on core/dbobject.class.php:
```diff
@@ -852,17 +855,13 @@ public function GetStrict($sAttCode)
 	}
 
     /**
-     * Get the value as it was before change with Set
```
I don't get why you removed the explanation

piRGoif replied:
I moved it to the @return tag.
Ok for you ?

bruno-ds reviewed (dismissed):
Beside the test recommended by Molkobain (and maybe a unit test if possible), 
It sound all good to me.

piRGoif committed:
:memo: Well, again some PHPDoc O:)

piRGoif committed:
:memo: Add more info in .doc README

piRGoif committed:
:memo: Replace inline @link by @see
@link are for URI, see https://docs.phpdoc.org/latest/references/phpdoc/inline-tags/link.html
